### PR TITLE
feat: notification preferences query

### PR DESCRIPTION
**Pull request by sshanzel**
As we introduced a notification preferences entity instead of just the muted ones, we should now support querying the said entity.

One concern though, the preferences here can grow in time, pulling all the preferences on every explicit subscription or muted can become a lot over time. Though on the initial rollout, there shouldn't be much to worry about (since we also filter by type) but just wanted to have it mentioned.

Thoughts are, we keep this or make it paginated though we will introduce another query to fetch a single preference based on the referenced id and the `notificationType` to identify the status and use it however the consuming end wants (verifying subscribed or muted already and reversing the action).

WT-1525 #done

### CODE REVIEW

**Comment by idoshamun** [comment]
@sshanzel what's the use case of querying them all?

**Comment by sshanzel** [comment]
> @sshanzel what's the use case of querying them all?

None yet at the initial release, but if we will ever present a list of notifications preferences, to provide options to users like removing muted selections, etc.

**Comment by sshanzel** [comment]
> @sshanzel if it's not needed I wouldn't proceed with this PR as it can potentially be a huge object as you mentioned. Is it part of the DR?

Not it was not, let's enforce the `type` requirement.

**Comment by idoshamun** [comment]
If it's not needed then I suggest to close the PR at this point.
There's no need to add additional functionality that is not required at the moment.
Let's be more aware next time 🙏

**Comment by sshanzel** [comment]
> If it's not needed then I suggest to close the PR at this point.
There's no need to add additional functionality that is not required at the moment.
Let's be more aware next time 🙏

Oh, this PR was required from the DR, just the requirement of having an optional `type` on the query.

**Comment by sshanzel** [comment]
Would love to hear what you guys think. 

cc @idoshamun @rebelchris @capJavert

**Review by rebelchris — COMMENTED** [review]
Nothing much on the code itself, but also concerned this might become very big.
At what point would we decide to query this?

As a once-off and then map to items on the frontend?
Alternative would be to only query the specific ID and get 1 result? (As it's only invoked on user action (options menu))

Sorry should have been tackled in the DR, but having some concerns with load of this massive query (potentially)

**Comment by sshanzel** [comment]
Should now be accurate with the updates from the DR.

File: __tests__/notifications.ts [hunk 1]
```diff
@@ -291,6 +302,118 @@ describe('query notifications', () => {
...
+  `;
+
+  it('should not authorize when not logged-in', () =>
+    testQueryErrorCode(
+      client,
+      { query: QUERY, variables: { data: [] } },
+      'UNAUTHENTICATED',
+    ));
+
+  const prepareNotificationPreferences = async () => {
```
Comment (rebelchris) [inline __tests__/notifications.ts:325]: I would suggest making this more extended for future proofing.

For instance:
- add multiple of the same type for different posts but same user
- also have on test that only queries one of the types and other should not auto return


Its just a failsafe for when code gets accidentally changed
Non blocking so will approve already

**Review by rebelchris — APPROVED** [review]
Nothing further from my side, just minor test suggestion

File: __tests__/notifications.ts (hunk 1 above)
Comment (sshanzel) [reply __tests__/notifications.ts:325]: > add multiple of the same type for different posts but same user

Sure. Let me add another one.

> also have on test that only queries one of the types and other should not auto return

We should have this covered as we already have other data for user `1` but with `source` type. First first test's return length was just one and verified to have the correct type (post).

Comment (sshanzel) [reply __tests__/notifications.ts:325]: Added! Thanks for the review! 🚀

File: src/schema/notifications.ts [hunk 2]
```diff
@@ -250,6 +316,17 @@ export const resolvers: IResolvers<any, Context> = traceResolvers({
...
+    notificationPreferences: (
+      _,
+      { data }: NotificationPreferenceInput,
+      { con, userId },
+    ): Promise<GQLNotificationPreference[]> => {
+      const params = data.reduce((args, value) => {
+        return [...args, { ...value, userId }];
+      }, []);
+
+      return con.getRepository(NotificationPreference).find({ where: params });
```
Comment (idoshamun) [inline src/schema/notifications.ts:328]: Please use graphorm. It's much safer.

Comment (sshanzel) [reply src/schema/notifications.ts]: Agreed. Should be updated now.

**Review by idoshamun — APPROVED** [review]

**Pull request merged**